### ISSUE TITLE
Caching for autoregressive decoding of Transformer

### ISSUE THREAD

**Issue by seewoo5** [pytorch/pytorch#49791]
## 🚀 Feature

[This blog post](https://tunz.kr/post/4) says that we can reduce the computation time of autoregressive decoding by using caching, since we do not have to re-compute the previous attentions. It seems that [OpenNMT](https://github.com/OpenNMT/OpenNMT-py/issues/1551) uses such caching, but I can't find such usage of caching in `nn.Transformer`. I wonder if such feature will be implemented and released in future.

## Motivation

Faster decoding.

## Pitch

Simple module for masked attention (with upper triangular mask) that enables caching for autoregressive decoding.

**Comment by mruberry** [comment]
Thank you for the suggestion, @seewoo5.

**Comment by speedcell4** [comment]
so, after three years, is there any progress on this?

**Comment by seewoo5** [comment]
> so, after three years, is there any progress on this?

I completely forgot about this, and also it's quite surprising that no one have worked on it yet.

**Comment by speedcell4** [comment]
> quite surprising that no one have worked on it yet

Yep, I am just wondering if the naive CUDA implementation of multihead attention from PyTorch makes training faster, otherwise, `huggingface/transformers` is obviously a better choice
